stage1/iottymux/iottymux.go

```diff
@@ -0,0 +1,601 @@
+// Copyright 2016 The rkt Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//+build linux
+
+package main
+
+import (
+	"bufio"
+	"encoding/json"
+	"flag"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net"
+	"os"
+	"path/filepath"
+	"strconv"
+	"syscall"
+	"time"
+
+	"github.com/appc/spec/schema/types"
+	"github.com/coreos/go-systemd/daemon"
+	rktlog "github.com/coreos/rkt/pkg/log"
+	stage1initcommon "github.com/coreos/rkt/stage1/init/common"
+	"github.com/kr/pty"
+)
+
+var (
+	log     *rktlog.Logger
+	diag    *rktlog.Logger
+	action  string
+	appName string
+	debug   bool
+)
+
+const (
+	// iottymux store several bits of information for a
+	// specific instance under /rkt/iottymux/<app>/
+	pathPrefix = "/rkt/iottymux"
+
+	// curren version of JSON API (for `list`)
+	apiVersion = 1
+)
+
+func init() {
+	// debug flag is not here, as it comes from env instead of CLI
+	flag.StringVar(&action, "action", "list", "Sub-action to perform")
+	flag.StringVar(&appName, "app", "", "Target application name")
+}
+
+// Endpoint represents a single attachable endpoint for an application
+type Endpoint struct {
+	// Name, freeform (eg. stdin, tty, etc.)
+	Name string `json:"name"`
+	// Domain, golang compatible (eg. tcp4, unix, etc.)
+	Domain string `json:"domain"`
+	// Address, golang compatible (eg. 127.0.0.1:3333, /tmp/file.sock, etc.)
+	Address string `json:"address"`
+}
+
+// Targets references all attachable endpoints, for status persistence
+type Targets struct {
+	Version int        `json:"version"`
```

Comment:
This JSON however is passed back to stage0 and to user  (in list mode). At some point it should unified with https://github.com/coreos/rkt/issues/3551.